File: C11/백준/Silver/15663. N과 M （9）/N과 M （9）.c
#include <stdio.h>
#include <stdlib.h>

int N, M;
int arr[8], visited[8];

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

void backtrack(int depth, int* currentSequence) {
    if (depth == M) {
        for (int i = 0; i < M; i++) {
            printf("%d ", currentSequence[i]);
        }
        printf("\n");
        return;
    }

    for (int i = 0; i < N; i++) {
        if (i > 0 && arr[i] == arr[i - 1] && !visited[i - 1]) {
            continue;
        }

        if (!visited[i]) {
            visited[i] = 1;
            currentSequence[depth] = arr[i];
            backtrack(depth + 1, currentSequence);
            visited[i] = 0;
        }
    }
}

int main() {
    scanf("%d %d", &N, &M);
    int arr_input[8];

    for (int i = 0; i < N; i++) {
        scanf("%d", &arr_input[i]);
    }

    for (int i = 0; i < N; i++) {
        arr[i] = arr_input[i];
    }

    qsort(arr, N, sizeof(int), compare);
    
    int currentSequence[8];
    backtrack(0, currentSequence);

    return 0;
}

File: Java/백준/Silver/15663. N과 M （9）/N과 M （9）.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[] arr;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] nm = br.readLine().split(" ");
        N = Integer.parseInt(nm[0]);
        M = Integer.parseInt(nm[1]);
        
        arr = Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray();
        
        Arrays.sort(arr);
        
        visited = new boolean[N];
        
        backtrack(0, new ArrayList<>());
    }

    public static void backtrack(int depth, List<Integer> currentSequence) {
        if (depth == M) {
            System.out.println(String.join(" ", currentSequence.stream()
                                                            .map(String::valueOf)
                                                            .toArray(String[]::new)));
            return;
        }

        for (int i = 0; i < N; i++) {
            if (i > 0 && arr[i] == arr[i - 1] && !visited[i - 1]) {
                continue;
            }

            if (!visited[i]) {
                visited[i] = true;
                currentSequence.add(arr[i]);
                backtrack(depth + 1, currentSequence);
                currentSequence.remove(currentSequence.size() - 1);
                visited[i] = false;
            }
        }
    }
}